### PR TITLE
HTTP20Adapter.close tests (supersedes #307)

### PR DESCRIPTION
**Pull request by KostyaEsmukov**
Add `close()` method implementation to the requests adapter and integration tests for it.

Supersedes #307
Related issue: #306

### CODE REVIEW

**Review by Lukasa — CHANGES_REQUESTED** [review]
Thanks for this! It would be good if we could add a test that demonstrates that calling `close` on the adapter both closes the underlying connections and also clears the connection cache. Mind adding those?

**Comment by KostyaEsmukov** [comment]
I've added that check to these two tests instead of creating a new one. Let me know if you think that this should be checked in a separate test.

File: test/test_integration.py [hunk 1]
```diff
@@ -1717,3 +1717,98 @@ def socket_handler(listener):
...
+        r = s.get('http://%s:%s' % (self.host, self.port))
+        connections_before_close = list(a.connections.values())
+        s.close()
+
+        # check that connections cache is empty
+        assert not a.connections
+
+        # check that all connections are actually closed
+        assert (connections_before_close and
+                all(conn._sock is None for conn in connections_before_close))
```
Comment (Lukasa) [inline test/test_integration.py]: Let's split the first assertion out and call it when we get the connections initially.

File: test/test_integration.py [hunk 2]
```diff
@@ -1717,3 +1717,98 @@ def socket_handler(listener):
...
+            s.mount('http://', a)
+            r = s.get('http://%s:%s' % (self.host, self.port))
+            connections_before_close = list(a.connections.values())
+
+        # check that connections cache is empty
+        assert not a.connections
+
+        # check that all connections are actually closed
+        assert (connections_before_close and
+                all(conn._sock is None for conn in connections_before_close))
```
Comment (Lukasa) [inline test/test_integration.py]: Same note here.

**Review by Lukasa — CHANGES_REQUESTED** [review]

**Review by Lukasa — APPROVED** [review]
Very nice, this looks really good. Thanks so much!

**Pull request merged**